File: src/types/architecture.ts
// Architecture model types — mirrors Rust types from reflection::architecture

export interface ComponentNode {
  id: string;
  component_path: string;
  component_type: string; // 'file' | 'module' | 'service'
  fix_count: number;
  error_count: number;
  causal_involvement_count: number;
  effective_fix_count: number;
  ineffective_fix_count: number;
  health_score: number;
  change_velocity: number;
  last_activity_at: string | null;
}

export interface ComponentEdge {
  source: string;
  target: string;
  relationship_type: string; // 'impacts' | 'co_changes_with'
  strength: number;
}

export interface GraphStats {
  total_components: number;
  total_relationships: number;
  avg_health_score: number;
  most_volatile: string[];
}

export interface ComponentGraph {
  nodes: ComponentNode[];
  edges: ComponentEdge[];
  stats: GraphStats;
}

export interface RebuildResult {
  components_count: number;
  relationships_count: number;
  workflow_name: string;
}

export interface FixSummary {
  id: string;
  fix_type: string;
  fix_description: string;
  effectiveness: string | null;
  applied_at: string | null;
}

export interface ImpactEntry {
  component_path: string;
  relationship_type: string;
  strength: number;
}

export interface ComponentDetails {
  node: ComponentNode;
  recent_fixes: FixSummary[];
  impacted_by: ImpactEntry[];
  impacts: ImpactEntry[];
}

export interface ImpactAnalysis {
  component: string;
  direct_impacts: ImpactEntry[];
  transitive_impacts: ImpactEntry[];
  total_impact_radius: number;
  highest_risk_path: string[];
}

// Temporal trend types (v103)

export interface TrendPoint {
  timestamp: string;
  value: number;
  count?: number;
}

export interface WorkflowTrends {
  workflow_name: string;
  convergence: TrendPoint[];
  fix_rate: TrendPoint[];
  velocity: TrendPoint[];
  total_fixes: TrendPoint[];
  snapshot_count: number;
}

export interface ComponentTrend {
  component_path: string;
  health_scores: TrendPoint[];
  fix_counts: TrendPoint[];
}

// Effectiveness over time types (v106)

export interface EffectivenessBucket {
  bucket: string;
  total: number;
  effective: number;
  ineffective: number;
  regression: number;
  effectiveness_rate: number;
}

export interface EffectivenessOverTime {
  workflow_name: string;
  bucket_type: string;
  buckets: EffectivenessBucket[];
}

File: src-tauri/src/reflection/architecture.rs
//! Architecture Model types.
//!
//! Aggregates component-level data from reflection fixes, causal events, and
//! knowledge into a queryable graph. The storage/query implementation lives
//! in `database/pg/reflection.rs`; this module only holds the shared types
//! and path normalization helpers used across the API surface.

use serde::Serialize;

// =============================================================================
// Types
// =============================================================================

#[derive(Debug, Clone, Serialize)]
pub struct ComponentNode {
    pub id: String,
    pub component_path: String,
    pub component_type: String,
    pub fix_count: u32,
    pub error_count: u32,
    pub causal_involvement_count: u32,
    pub effective_fix_count: u32,
    pub ineffective_fix_count: u32,
    pub health_score: f64,
    pub change_velocity: f64,
    pub last_activity_at: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ComponentEdge {
    pub source: String,
    pub target: String,
    pub relationship_type: String,
    pub strength: u32,
}

#[derive(Debug, Clone, Serialize)]
pub struct GraphStats {
    pub total_components: u32,
    pub total_relationships: u32,
    pub avg_health_score: f64,
    pub most_volatile: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ComponentGraph {
    pub nodes: Vec<ComponentNode>,
    pub edges: Vec<ComponentEdge>,
    pub stats: GraphStats,
}

#[derive(Debug, Clone, Serialize)]
pub struct RebuildResult {
    pub components_count: u32,
    pub relationships_count: u32,
    pub workflow_name: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct FixSummary {
    pub id: String,
    pub fix_type: String,
    pub fix_description: String,
    pub effectiveness: Option<String>,
    pub applied_at: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ImpactEntry {
    pub component_path: String,
    pub relationship_type: String,
    pub strength: u32,
}

#[derive(Debug, Clone, Serialize)]
pub struct ComponentDetails {
    pub node: ComponentNode,
    pub recent_fixes: Vec<FixSummary>,
    pub impacted_by: Vec<ImpactEntry>,
    pub impacts: Vec<ImpactEntry>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ImpactAnalysis {
    pub component: String,
    pub direct_impacts: Vec<ImpactEntry>,
    pub transitive_impacts: Vec<ImpactEntry>,
    pub total_impact_radius: u32,
    pub highest_risk_path: Vec<String>,
}

// =============================================================================
// Path Normalization
// =============================================================================

/// Normalize a raw file path: lowercase, forward slashes, strip `./` prefix and drive letters.
pub fn normalize_component_path(raw: &str) -> String {
    let mut path = raw.replace('\\', "/").to_lowercase();
    // Strip drive letters like C:/
    if path.len() >= 3 && path.as_bytes()[1] == b':' && path.as_bytes()[2] == b'/' {
        path = path[3..].to_string();
    }
    // Strip leading ./
    if let Some(stripped) = path.strip_prefix("./") {
        path = stripped.to_string();
    }
    path
}

/// Infer component type from a path.
pub fn infer_component_type(path: &str) -> &str {
    if path.contains("/services/") || path.contains("/service/") {
        return "service";
    }
    if path.ends_with('/') || !path.contains('.') {
        return "module";
    }
    "file"
}
